File: 2018/advent-of-code/19.pl
#!/usr/bin/perl
## CHEAT!: https://arstechnica.com/gaming/2018/12/for-the-first-time-ever-disney-posts-a-pixar-short-on-youtube-for-free/
use warnings;
use strict;
use autodie;
use 5.020;

my $ip_reg;
my @program;

my %ops =
  (
   addr => sub { "registers[$_[2]] = registers[$_[0]] + registers[$_[1]];" },
   addi => sub { "registers[$_[2]] = registers[$_[0]] + $_[1];" },
   mulr => sub { "registers[$_[2]] = registers[$_[0]] * registers[$_[1]];" },
   muli => sub { "registers[$_[2]] = registers[$_[0]] * $_[1];" },
   banr => sub { "registers[$_[2]] = registers[$_[0]] & registers[$_[1]];" },
   bani => sub { "registers[$_[2]] = registers[$_[0]] & $_[1];" },
   borr => sub { "registers[$_[2]] = registers[$_[0]] | registers[$_[1]];" },
   bori => sub { "registers[$_[2]] = registers[$_[0]] | $_[1];" },
   setr => sub { "registers[$_[2]] = registers[$_[0]];" },
   seti => sub { "registers[$_[2]] = $_[0];" },
   gtir => sub { "registers[$_[2]] = $_[0] > registers[$_[1]];" },
   gtri => sub { "registers[$_[2]] = registers[$_[0]] > $_[1];" },
   gtrr => sub { "registers[$_[2]] = registers[$_[0]] > registers[$_[1]];" },
   eqir => sub { "registers[$_[2]] = $_[0] == registers[$_[1]];" },
   eqri => sub { "registers[$_[2]] = registers[$_[0]] == $_[1];" },
   eqrr => sub { "registers[$_[2]] = registers[$_[0]] == registers[$_[1]];" }
  );

while (<>) {
  if (/#ip (\d+)/) {
    $ip_reg = $1;
  } elsif (/(\w+) (\d+) (\d+) (\d+)/) {
    push @program, $ops{$1}->($2, $3, $4);
  }
}

say "Creating day19.c";
open my $src, ">", "day19.c";

print $src <<EOC;
#include <stdio.h>

int registers[6];
void run(void);

void zero_registers(void) {
  for (int n = 0; n < 6; n += 1) { registers[n] = 0; }
}

int main(void) {
  zero_registers();
  run();
  printf("Part 1: %d\\n", registers[0]);
  zero_registers();
  registers[0] = 1;
  run();
  printf("Part 2: %d\\n", registers[0]);
  return 0;
}

EOC

my $ninsns = @program;
  print $src <<EOC;
void run(void) {
  int ip = 0;
  void *labels[] = {
EOC
say $src "    &&instr_$_," for 0 .. $ninsns - 1;
say $src "  };";
for (0 .. $ninsns - 1) {
  say $src "  instr_$_:";
  my $uses_ip = $program[$_] =~ m/registers\[$ip_reg\]/;
  my $sets_ip = $program[$_] =~ m/registers\[$ip_reg\] =/;
  say $src "  registers[$ip_reg] = ip;" if $uses_ip;
  say $src "  ", $program[$_];
  if ($sets_ip) {
    say $src "  ip = registers[$ip_reg] + 1;";
  } else {
    say $src "  ip += 1;";
  }
  say $src "  if (ip >= $ninsns) { return; }";
  say $src "  goto *labels[ip];";
}
say $src "}";
close $src;

say "Compiling day19.c";
system "gcc -std=gnu11 -O3 -march=native -Wall -Wextra -o day19 day19.c";
say "Running day19 executable";
system "./day19";

File: 2018/advent-of-code/19.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import argparse

class VM(object):
    def __init__(self, ip):
        self.ip = ip
        self.reg = [0, 0, 0, 0, 0, 0]
        self.reg[self.ip] -= 1

    def run(self, prog):
        while -1 <= self.reg[self.ip] < len(prog)-1:
            self.reg[self.ip] += 1
            op, a, b, c = prog[self.reg[self.ip]]
            VM.__dict__[op](self.reg, a, b, c)
            print(self.reg)
        return self.reg

    def addr(R, a, b, c): R[c] = R[a] + R[b]
    def addi(R, a, b, c): R[c] = R[a] + b

    def mulr(R, a, b, c): R[c] = R[a] * R[b]
    def muli(R, a, b, c): R[c] = R[a] * b

    def banr(R, a, b, c): R[c] = R[a] & R[b]
    def bani(R, a, b, c): R[c] = R[a] & b

    def borr(R, a, b, c): R[c] = R[a] | R[b]
    def bori(R, a, b, c): R[c] = R[a] | b

    def setr(R, a, b, c): R[c] = R[a]
    def seti(R, a, b, c): R[c] = a

    def gtir(R, a, b, c): R[c] = 1 if a > R[b] else 0
    def gtri(R, a, b, c): R[c] = 1 if R[a] > b else 0
    def gtrr(R, a, b, c): R[c] = 1 if R[a] > R[b] else 0

    def eqir(R, a, b, c): R[c] = 1 if a == R[b] else 0
    def eqri(R, a, b, c): R[c] = 1 if R[a] == b else 0
    def eqrr(R, a, b, c): R[c] = 1 if R[a] == R[b] else 0


def MAIN(argv):
    with open(argv.fname, 'r', encoding='UTF-8') as fh:
        prog = []
        for line in fh.readlines():
            if line.startswith("#ip "):
                vm = VM(int(line[4]))
            else:
                l = line.split()
                prog.append([ l[0], int(l[1]), int(l[2]), int(l[3]) ])

#     print("Part 1:", vm.run(prog))

    vm.reg = [1, 0, 0, 0, 0, 0]
    vm.reg[vm.ip] -= 1
    print("Part 2:", vm.run(prog))

def getopts():
    parser = argparse.ArgumentParser(description="""Advent of code day 19""")
    parser.add_argument('fname', type=str, nargs='?', default="19.in", help='File name')
    return parser.parse_args()

if __name__ == '__main__':
    MAIN(getopts())
